### PR TITLE
Store tx on localstorage

### PR DESCRIPTION
**Pull request by antomor**
## What

- Store the transactions on local storage to retrieve them on app reload
- merge the locally stored transactions with the ones retrieved from the contract and the server.

## Why

- we can reload the application without losing the transactions created (and not yet approved)
- we can store the transactions even before publishing them
- we can keep track of the signer the signed the transaction off-chain

### CODE REVIEW

File: src/helpers/missingTypes.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { SafeSignature } from '@gnosis.pm/safe-core-sdk/'
+
+// The package '@gnosis.pm/safe-service-client' doesn't export the type SafeMultisigConfirmationResponse
```
Comment (ilanolkies) [inline src/helpers/missingTypes.ts:3]: Should we rise an issue in their repo?

**Review by ilanolkies — COMMENTED** [review]
Works for me! @jessgusclark can check code perspective.

File: src/helpers/missingTypes.ts (hunk 1 above)
Comment (antomor) [reply src/helpers/missingTypes.ts:3]: Yep, I think so

**Comment by jessgusclark** [comment]
An issue with checksums (perhaps) and new safes

1. create a new safe
2. Fund the safe with the faucet
3. Create a transaction
4. refresh page.
5. transaction does not show up in UI, but is in localstorage:

Screen Shot 2021-07-21 at 1 11 39 PM

**Comment by jessgusclark** [comment]
An issue with a rejection transaction that is not executed showing up as pending after refresh

1. existing safe from above
2. create a transfer transaction
3. create a rejection transaction
4. sign the transfer transaction off-chain
5. refresh (probably not needed but I wanted to check both transactions were saved)
6. execute the transfer transaction
7. see the two transactions sorted correctly into their tabs/buckets: executed and rejected
8. Refresh site
9. Rejection transaction shows up under "Pending Transactions" with a nonce of X even though the safe's nonce should be x+1

**Review by jessgusclark — CHANGES_REQUESTED** [review]
The code looks great and THANK YOU for fixing my spelling mistakes! I left two comments regarding issues that I found with the UX.

**Comment by antomor** [comment]
> An issue with checksums (perhaps) and new safes
> 
>     1. create a new safe
> 
>     2. Fund the safe with the faucet
> 
>     3. Create a transaction
> 
>     4. refresh page.
> 
>     5. transaction does not show up in UI, but is in localstorage:
> 
> 
> Screen Shot 2021-07-21 at 1 11 39 PM

Thank you for pointing it out 🙏🏼  
I confirm that the issue is related to the safe address case.

**Comment by antomor** [comment]
> An issue with a rejection transaction that is not executed showing up as pending after refresh
> 
>     1. existing safe from above
> 
>     2. create a transfer transaction
> 
>     3. create a rejection transaction
> 
>     4. sign the transfer transaction off-chain
> 
>     5. refresh (probably not needed but I wanted to check both transactions were saved)
> 
>     6. execute the transfer transaction
> 
>     7. see the two transactions sorted correctly into their tabs/buckets: executed and rejected
> 
>     8. Refresh site
> 
>     9. Rejection transaction shows up under "Pending Transactions" with a nonce of X even though the safe's nonce should be x+1

Although this particular use case is solved, I think a better solution would be to publish all the transactions, but it's not possible with the current version of the `safe-transaction-service` client.

**Comment by ilanolkies** [comment]
Maybe the UX is better if we add a new tab for the “not submitted” transactions. Instead of merging local storage and tx service, we can do one list for each. The “not submitted” ones would allow to sign on chain or off chain, and submit.

**Comment by jessgusclark** [comment]
@ilanolkies @antomor 

Sorry, I don't understand the issue with publishing two transactions with the same nonce to the service. I thought this was possible.

Nonetheless, How about a different scenario:

1. create a transaction
2. sign it off-chain
3. publish it to the transaction service
4. refresh
5. create rejection transaction
6. sign rejection transaction
7. execute rejection transaction
4. refresh site
5. the "send value" transaction shows up twice under "Rejected" and the executed [rejected] transaction is missing. The nonce is set to 1 so a transaction with the nonce of 0 was executed.


When retrieving the transactions from the service or localStorage, the UX should check if the nonce of the transaction is less than the safe's nonce. If true, and the transaction is not executed, then the transaction is "rejected".

**Comment by antomor** [comment]
> @ilanolkies @antomor
> 
> Sorry, I don't understand the issue with publishing two transactions with the same nonce to the service. I thought this was possible.

The problem here was related to the `safe-transaction-service` client, that requires the signature as mandatory parameter.
We can circumvent that issue by performing the API request directly, without using the client.

> 
> Nonetheless, How about a different scenario:
> 
>     1. create a transaction
> 
>     2. sign it off-chain
> 
>     3. publish it to the transaction service
> 
>     4. refresh
> 
>     5. create rejection transaction
> 
>     6. sign rejection transaction
> 
>     7. execute rejection transaction
> 
>     8. refresh site
> 
>     9. the "send value" transaction shows up twice under "Rejected" and the executed [rejected] transaction is missing. The nonce is set to 1 so a transaction with the nonce of 0 was executed.
> 
> 
> When retrieving the transactions from the service or localStorage, the UX should check if the nonce of the transaction is less than the safe's nonce. If true, and the transaction is not executed, then the transaction is "rejected".

Although publishing the transaction does solve a lot of misalignment issues, this situation described by @jessgusclark can still occur. This is due to the fact that the service requires some time before updating the rejection transaction. 
So just after the transaction is executed, there will be two transactions in an undefined state:
- they look like they are pending, but they are not since that the safe nonce is greater
- they could be set as rejected, but actually, one of them is executed, and only the other one is rejected.
The situation is solved just after the server receives the update from the network (with the `trace_block` method).

**Comment by antomor** [comment]
Since that we changed completely strategy and instead of storing transactions locally, we store them on the service, I'm replacing this PR with a new one: https://github.com/rsksmart/multisig-sample-app/pull/36

**Pull request closed without merging**